cpp/src/fil/treelite_import.cu

```diff
@@ -69,32 +69,73 @@ int tree_root(const tl::Tree<T, L>& tree)
   return 0;  // Treelite format assumes that the root is 0
 }
 
-template <typename T, typename L>
-inline int max_depth(const tl::Tree<T, L>& tree)
+// a no-op placeholder for values and callables alike
+struct empty {
+  template <typename... Args>
+  void operator()(Args...)
+  {
+  }
+};
+
+/** walk a Treelite tree, visiting each inner node with visit_inner and each leaf node with
+  visit_leaf. See walk_tree::element::state documentation for how TraversalState is retained
+during traversal. Any per-tree state during traversal should be captured by the lambdas themselves.
+  visit_inner(int node_id, TraversalState state) should return a pair of new states, one for
+each child node. visit_leaf(int, TraversalState) returns nothing.
+**/
+template <typename T, typename L, typename InnerFunc, typename LeafFunc = empty>
+inline void walk_tree(const tl::Tree<T, L>& tree,
+                      InnerFunc visit_inner,
+                      LeafFunc visit_leaf = empty())
 {
-  // trees of this depth aren't used, so it most likely means bad input data,
-  // e.g. cycles in the forest
-  const int DEPTH_LIMIT = 500;
-  int root_index        = tree_root(tree);
-  typedef std::pair<int, int> pair_t;
-  std::stack<pair_t> stack;
-  stack.push(pair_t(root_index, 0));
-  int max_depth = 0;
-  while (!stack.empty()) {
-    const pair_t& pair = stack.top();
-    int node_id        = pair.first;
-    int depth          = pair.second;
-    stack.pop();
-    while (!tree.IsLeaf(node_id)) {
-      stack.push(pair_t(tree.LeftChild(node_id), depth + 1));
-      node_id = tree.RightChild(node_id);
-      depth++;
-      ASSERT(depth < DEPTH_LIMIT, "depth limit reached, might be a cycle in the tree");
+  if constexpr (std::is_invocable<InnerFunc, int>()) {
+    /// wrapper for empty path state
+    walk_tree(
+      tree,
+      [&](int nid, empty val) {
+        visit_inner(nid);
+        return std::pair<empty, empty>();
+      },
+      [&](int nid, empty val) { visit_leaf(nid); });
+  } else {
+    using TraversalState = decltype(visit_inner(int(), {}).first);
+    /// needed to visit a node
+    struct element {
+      int tl_node_id;
+      /// Retained while visiting nodes on a single path from root to leaf.
+      /// This generalizes the node index that's carried over during inference tree traversal.
+      TraversalState state;
+    };
+    std::stack<element> stack;
+    stack.push(element{tree_root(tree), TraversalState()});
```

Comment:
```suggestion
    stack.emplace(tree_root(tree), TraversalState());
```

Reply:
Does C++ allow that when an explicit constructor wasn't defined?
nvcc didn't accept this.
[GodBolt reproducers](https://godbolt.org/z/Ms8Gcvxzb)